packages/opentelemetry-resources/src/platform/node/detectors/AwsEcsDetector.ts

```diff
@@ -0,0 +1,97 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Resource } from '../../../Resource';
+import { CONTAINER_RESOURCE } from '../../../constants';
+import { Detector } from '../../../types';
+import { ResourceDetectionConfigWithLogger } from '../../../config';
+import * as util from 'util';
+import * as fs from 'fs';
+import * as os from 'os';
+
+export const cache = {
+  readFileAsync: util.promisify(fs.readFile),
+};
+
+/**
+ * The AwsEcsDetector can be used to detect if a process is running in AWS
+ * ECS and return a {@link Resource} populated with data about the ECS
+ * plugins of AWS X-Ray. Returns an empty Resource if detection fails.
+ */
+class AwsEcsDetector implements Detector {
+  readonly CONTAINER_ID_LENGTH = 64;
+  readonly DEFAULT_CGROUP_PATH = '/proc/self/cgroup';
+  readonly ECS_METADATA_KEY_V4 = 'ECS_CONTAINER_METADATA_URI_V4';
+  readonly ECS_METADATA_KEY_V3 = 'ECS_CONTAINER_METADATA_URI';
+
+  async detect(config: ResourceDetectionConfigWithLogger): Promise<Resource> {
+    if (
+      !process.env.ECS_CONTAINER_METADATA_URI_V4 &&
+      !process.env.ECS_CONTAINER_METADATA_URI
+    ) {
+      config.logger.debug('AwsEcsDetector failed: Process is not on ECS');
+      return Resource.empty();
+    }
+
+    let hostName, containerId;
+    try {
+      hostName = os.hostname();
+    } catch (e) {
+      config.logger.warn(
+        `AwsEcsDetector failed to read host name: ${e.message}`
+      );
+    }
+    try {
+      const rawData = await cache.readFileAsync(
+        this.DEFAULT_CGROUP_PATH,
+        'utf8'
+      );
+      containerId = this._getContainerId(rawData);
+    } catch (e) {
+      config.logger.warn(
+        `AwsEcsDetector failed to read container ID: ${e.message}`
+      );
+    }
+
+    return !hostName && !containerId
+      ? Resource.empty()
+      : new Resource({
+          [CONTAINER_RESOURCE.NAME]: hostName || '',
+          [CONTAINER_RESOURCE.ID]: containerId || '',
+        });
+  }
+  /**
+   * Read container ID from cgroup file
+   * In ECS, even if we fail to find target file
+   * or target file does not contain container ID
+   * we do not throw an error but throw warning message
+   * and then return null string
+   */
+  private _getContainerId(rawData: string): string {
+    const splitData = rawData.trim().split('\n');
+
+    for (let i = 0; i < splitData.length; i += 1) {
+      if (splitData[i].length > this.CONTAINER_ID_LENGTH) {
+        return splitData[i].substring(
+          splitData[i].length - this.CONTAINER_ID_LENGTH
+        );
+      }
+    }
+    return '';
```

Comment:
I would suggest when doing loops to use the `for..of` syntax

```suggestion
    for (const line of splitData) {
      if (line.length > this.CONTAINER_ID_LENGTH) {
        return line.substring(
          line.length - this.CONTAINER_ID_LENGTH
        );
      }
    }
    return '';
```

Reply:
Sorry I initially had a typo and didnt work. I have changed and made a new commit. Thank you for your help!